lib/puppet/configurer.rb

```diff
@@ -246,9 +266,16 @@ def run_internal(options)
   private :run_internal
 
   def send_report(report)
+    report.finalize_report
+    report_send_time = thinmark {
+      Puppet::Transaction::Report.indirection.save(report, nil, :environment => Puppet::Node::Environment.remote(@environment)) if Puppet[:report]
+    }
+    # Includes the indirection save time in the last_run_summary.yaml
+    # and in the summary
+    report.add_times(:reporting, report_send_time)
+    report.finalize_report
     puts report.summary if Puppet[:summarize]
     save_last_run_summary(report)
-    Puppet::Transaction::Report.indirection.save(report, nil, :environment => Puppet::Node::Environment.remote(@environment)) if Puppet[:report]
```

Comment:
I'm a little concerned about these changes, because this area of the code is pretty fragile, and breakage can have big consequences, submitted reports are missing info needed for console, compliance, etc.

This change means the report summary data (`report.raw_summary`) that we submitted is different than the last run report saved on the agent. I'd prefer that the report be immutable once it's submitted.

Is it safe to call `finalize_report` multiple times? This change means `finalize_report` will be called twice before we send the report, and a third time prior to saving the last run summary.

Previously `save_last_run_summary` & `report.summary` came before `Report.indirection.save`, and now the order is reversed. Does the call to `Report.indirection.save` rely on side effects caused by the former?

Given these questions I have to ask is there another way to collect the `report_send_time` that doesn't involve modifying the report? I understand why it's desirable to have, I'm just concerned about what it might break.